importccl: move to pkg/sql/importer

dt opened the pull request.

dt committed:
backupccl: move RowCount to roachpb

Release note: none.

dt committed:
jobspb,backupccl: move DescriptorRewrite out of RestoreDetails

DescriptorRewrite is used in IMPORT as well as RESTORE so moving it out of RestoreDetails.
Also adding the alias for DescRewriteMap to jobspb so that importccl does not need to depend
on backupccl to get it.

Release note: none.

dt committed:
backupccl: move descriptor privilege setting to catalog/ingesting

Release note: none.

dt committed:
backupccl: move WriteDescriptors to catalog/ingesting

This is the last non-test backupccl function that importccl was using.

Release note: none.

dt committed:
backupccl: move CheckEmittedEvents to jobstest pkg

Release note: none.

dt committed:
storageccl: move ingestsize setting to kv/bulk

Release note: none.

dt committed:
importccl: split ccl-only tests into own package

This moves the tests from improtccl that depend on CCL deatures (tenants, multi-region, etc)
into their own separate package that can depend on importccl, soon to be renamed sql/importer,
so that the remaining importccl tests do not require ccl features.

Release note: none.

dt committed:
utilccl: move ExecutionResultHeaders to pkg/jobs

Release note: none.

dt committed:
utilccl: move IsPermanentBulkJobError to pkg/jobs/joberror

Release note: none.

dt committed:
importccl: inline workloadccl/sstable.ToTableDesc

This is the only user.

Release note: none.

dt committed:
importccl: move to pkg/sql/importer

Release note: none.

dt committed:
sql/catalog/rewrite: de-stutter rewrite functions

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76832)
<!-- Reviewable:end -->

dt committed:
server: inject sql/importer without ccl hooks

Release note: none.

dt commented:
Unfortunately I was like 85% done with this before I realized I still had my editor configured to use the more strict mvdan/gofumpt on every save rather than the usual gofmt / crlfmt, so there are some spurious / unrelated diffs in files I touched. They're all still crlfmt-compliant so they're not bad, and generally in fact I think it's rules are good which I why I usually use it, but it just adds some noise to a big change that I didn't intend. Getting them back out though it merge tedious (since, again, gofmt/crlfmt doesn't think they're wrong or change them) so unless there's an objection I was just going to go with it. The diffs are mostly things like trailing `}`  on its own line for multi-line struct literals or whatever.

dt commented:
For now I haven't moved the `_` import to run the plan and job registration hooks from pkg/ccl to pkg/server yet just since at the moment that creates a cycle for the tests that are still in pkg/sql/importer but depend on pkg/server via testcluster. We'll need to do a pass where we split the testcluster-using tests out into importer_test before we can move the injection import to pkg/server, but I wanted to stop here before this PR got too much bigger.

dt committed:
pkg/sql/importer: switch tests to use serverutils.StartNewTestCluster

Remove the direct dependency on testcluster from all non-_test pacakge tests.

Release note: none.

dt commented:
Okay, I think I fixed all the tests that weren't in `importer_test` to stop depending on testcluster and server directly and use the shim, so I was able to remove the ccl-side hook in favor of server now, so pure OSS binary should be able to run IMPORT as well.

dt commented:
for the most part this is all mechanical code movement of whole files or whole functions as-is.

There were a couple though that required a little editing to work which might warrant closer review, namely:

- `move ingestsize setting to kv/bulk`: I removed some 'let's guess at the raft command overhead' that had no business being client-side. I almost got ride of the 'min-of-command-size-or-setting' logic entirely and may still say we should do so in a follow-up. You should just a) not change that setting b) keep it to values kv won't reject if you do, but we can do that in a targeted change.
- `move descriptor privilege setting to catalog/ingesting`: This required a little change to how we keep the temp system db perms handled one way and everything else another, changing how `wroteDbs` is populated.

dt commented:
We could probably de-hook this planning entirely if we wanted at some point now but that seems like it can wait -- I mostly just wanted to get the code moved for now before the 22.1 branch is cut to minimize backport conflicts over the upcoming cycle.

adityamaru reviewed (approved):
Did one pass through all the commits, and it LGTM. Maybe we should get another stamp on this though considering the blast radius.

dt commented on pkg/jobs/jobstest/traceutils.go:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 The Cockroach Authors.
```
Ah, I thought the emitted events being checked here were trace events? but they're event log events, aren't they?

dt replied:
renamed to logutils.go

dt commented on pkg/ccl/importccl/import_job.go:
```diff
@@ -564,9 +564,9 @@ func prepareNewTablesForIngestion(
+	if err := ingesting.WriteDescriptors(ctx, p.ExecCfg().Codec, txn, p.User(), descsCol,
 		nil /* databases */, nil, /* schemas */
-		tableDescs, nil, tree.RequestedDescriptors, seqValKVs); err != nil {
+		tableDescs, nil, tree.RequestedDescriptors, seqValKVs, "" /* exemptParentName */); err != nil {
```
these comments: still the worst.

dt commented on pkg/ccl/importerccl/ccl_test.go:
```diff
@@ -0,0 +1,470 @@
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
```
I _think_ the pattern is somewhat established, but added a comment.

dt commented:
TFTR!

I'll wait until tomorrow to see if I get any other reviewers by then before I go OOO, but will probably aim to merge this one sooner rather than later with how conflict prone it is; while the files affected are numerous, I hope the actual blast radius is pretty small given it is mostly just movement and not changing behaviors. 

I'm sure we could get more reviewers to bikeshed some of the naming of the new packages, but I'd sorta rather do any of that in targeted, single-package follow-ups, once this one is merged with the actual untangling work it needed to do at once behind us.

adityamaru commented:
SGTM, thanks!

dt commented:
TFTR!

Checked with schema offline about the overall idea and what bits we were moving to catalog here and we’re broadly aligned on the bits that moved her moving and can refine more later.

let’s see if it can get by bors now.

bors r+

adityamaru commented:
Could `single on` it to prevent it batching with other PRs.

dt commented:
bors r+

bors single on
